### PR TITLE
Optimised flow to submit API events to Pulsar

### PR DESCRIPTION
**Pull request by severinson**
This PR adds code to efficiently submit several API events in bulk to Pulsar.
In particular, it makes bulk job cancellation much faster.

### CODE REVIEW

File: internal/armada/server/submit_to_log.go [hunk 1]
```diff
@@ -587,538 +625,3 @@ func (srv *PulsarSubmitServer) SubmitApiEvent(ctx context.Context, apiEvent *api
 
 	return nil
 }
-
-// PulsarSequenceFromApiEvent converts an api.EventMessage into the corresponding Pulsar event
-// and returns an EventSequence containing this single event.
-func PulsarSequenceFromApiEvent(msg *api.EventMessage) (sequence *armadaevents.EventSequence, err error) {
```
Comment (severinson) [inline internal/armada/server/submit_to_log.go:593]: This function is moved into eventutil.

File: internal/common/eventutil/eventutil.go [hunk 2]
```diff
@@ -363,3 +363,591 @@ func K8sObjectMetaFromLogObjectMeta(meta *armadaevents.ObjectMeta) *metav1.Objec
...
+		sequences = append(sequences, sequence)
+	}
+	return sequences
+}
+
+// EventSequenceFromApiEvent converts an api.EventMessage into the corresponding Pulsar event
+// and returns an EventSequence containing this single event.
+// We map API events to sequences one-to-one because each API event may contain different (queue, jobSet, userId),
+// which must be common to all events in a sequence.
+func EventSequenceFromApiEvent(msg *api.EventMessage) (sequence *armadaevents.EventSequence, err error) {
```
Comment (severinson) [inline internal/common/eventutil/eventutil.go:419]: No changes made to this function. Just moved.

File: internal/armada/server/submit_to_log.go [hunk 3]
```diff
@@ -551,9 +549,70 @@ func (srv *PulsarSubmitServer) GetQueueInfo(ctx context.Context, req *api.QueueI
...
+	if err != nil {
+		return err
+	}
+
+	// Collect any errors experienced by the async send and return.
+	var result *multierror.Error
+	for range sequences {
+		result = multierror.Append(result, <-ch)
+	}
+	return result.ErrorOrNil()
```
Comment (d80tb7) [inline internal/armada/server/submit_to_log.go:610]: we could spin here until we've published all the messages.

File: internal/common/eventutil/eventutil.go [hunk 4]
```diff
@@ -363,3 +363,591 @@ func K8sObjectMetaFromLogObjectMeta(meta *armadaevents.ObjectMeta) *metav1.Objec
...
+	// Reduce the sequences to the smallest number possible.
+	return CompactEventSequences(sequences), nil
+}
+
+// CompactEventSequences converts a []*armadaevents.EventSequence into a []*armadaevents.EventSequence of minimal length.
+// In particular, it moves all events part of sequences with equal (queue, jobSetName, userId) into a single sequence.
+// The groups field is set to that of the first event in the sequence for a each (queue, jobSetName, userId).
+// This function may mutates the input sequences, i.e., it should be used like
+// sequences = CompactEventSequences(sequences).
+func CompactEventSequences(sequences []*armadaevents.EventSequence) []*armadaevents.EventSequence {
```
Comment (d80tb7) [inline internal/common/eventutil/eventutil.go:389]: maybe add a max batch size here/

**Review by d80tb7 — APPROVED** [review]

**Pull request merged**